Created PPP interface and PPP service classes to netsocket

mikaleppanen opened the pull request:
### Description

Created PPP interface and PPP service classes to netsocket. This allows that both lwip and nanostack can use a common PPP service for cellular connectivity. Lwip PPP protocol is separated from lwip and moved under netsocket as an independent entity. PPP offers much similar interface for (onboard) network stack as the EMAC and L3IP. PPPInterface class has been created as similar helper class as the EMACInterface and L3IPInterface are.

The "nsapi_ppp.h" interface is kept same, and the implementation for it is in ppp service file "ppp_nsapi.cpp". Cellular classes using the "nsapi_ppp.h" interface should not need updates. Existing ppp configuration under lwip mbed_lib.json is still supported, although configuration under ppp service should be used instead.

Most major changes to internal structure of PPP were, to replace its interface to lwip netif with new one and to make PPP to use network stack's memory manager for memory allocations, instead of direct access to lwip memory buffers.

### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers



### Release Notes

Add Cellular PPP connection support to Nanostack. Support is made by creating a common PPP service class that both lwIP and Nanostack can use for cellular connectivity.

Affected components are lwIP and Nanostack on board IP stacks, new netsocket PPP interface and PPP service class. Cellular network interface, cellular classes and "nsapi_ppp.h" interface are not modified. No changes are required from the users of the Cellular network interface to continue using the lwIP as the default IP stack.

To use the Nanostack IP stack with the Cellular network interface instead of the lwIP stack, cellular application configuration must be modified in following way:

Set the default stack to nanostack

 `"nsapi.default-stack": "NANOSTACK"`

Configure PPP to IPv6 only mode (Nanostack does not support IPv4)

`"ppp.ipv4-enabled": false,
 "ppp.ipv6-enabled": true,`

KariHaapalehto committed:
Moved PPP from lwip directory to under netsocket

Moved PPP and renamed files and constants.

Commit:
Created PPP service class

Created PPP service class that encapsulates the PPP protocol.
Class is similar to EMAC and L3IP classes with additional methods
to read IP and DNS server addresses negotiation using PPP and
to set PPP specific parameters (file handle for modem access etc.).

PPP service can use on its own thread or in run in mbed os event
Queue thread.

Added ppp_nsapi.cpp module that implements the nsapi_ppp.h
services.

Added ppp_nsapi.cpp module that implements the nsapi_ppp.h
services.

Commit:
Created netsocket classes for PPP service and PPP interface

PPP service encapsulates the PPP protocol. PPP interface can be used as
helper class to bind PPP protocol with network stack (similar to
EMAC and L3IP interface). Added PPP interface to onboard network
stack class.

Commit:
lwIP changes for PPP service

Created (a new) PPP interface for PPP service. Removed lwip
dependencies to PPP (memory allocations etc.). Moved PPP
configuration options away from lwIP mbed_lib.json to new
PPP service. For backwards compatibility, using the old
options is also currently supported.

Commit:
Nanostack changes for PPP service

Created PPP interface for PPP service. Re-used the ethernet tasklet
and PHY driver structure for PPP.

Commit:
Removed pbuf operations that were commented out

Commit:
Increased PPP stack size based on testing

Increased stack size from 768 to 816.

kivaisan commented on features/nanostack/mbed-mesh-api/source/NanostackPPPInterface.cpp:
```diff
@@ -0,0 +1,402 @@
+    if (address_type != PHY_MAC_64BIT) {
+        return -1;
+    }
+    //memcpy(unique_id, address_ptr, 8);
```
Are these commented codes needed?

mikaleppanen replied:
Removed the write function since not valid for ppp.

kivaisan commented on features/nanostack/mbed-mesh-api/source/NanostackPPPInterface.cpp:
```diff
@@ -0,0 +1,402 @@
+        phy.PHY_MAC = iid64;
+        phy.address_write = ppp_phy_address_write;
+        phy.driver_description = const_cast<char *>("PPP");
+        phy.link_type = static_cast<phy_link_type_e>(5); // PHY_LINK_PPP To compile without nanostack changes
```
Is there some bigger reason not to define `PHY_LINK_PPP` in `phy_link_type_e`?

mikaleppanen replied:
Added the constant to phy_link_type_e.

kivaisan commented on features/netsocket/PPP.h:
```diff
@@ -0,0 +1,165 @@
+     * @param uname User name
+     * @param passwrd Password
+     */
+    virtual void set_uname_password(const char *uname, const char *password) = 0;
```
Cellular has used `set_credentials` name for these.

mikaleppanen replied:
Renamed.

kivaisan commented on features/netsocket/ppp/ppp.c:
```diff
@@ -927,8 +905,8 @@ void ppp_input(ppp_pcb *pcb, struct pbuf *pb) {
         if (protp->protocol == protocol) {
-          pb = pbuf_coalesce(pb, PBUF_RAW);
-          (*protp->input)(pcb, (u8_t*)pb->payload, pb->len);
+          ///////// pb = pbuf_coalesce(pb, PBUF_RAW);
```
Why commented out?

mikaleppanen replied:
Removed as not needed (always a single buffer on this stage).

Commit:
lwIP adds now PPP DNS servers to default interface

If PPP interface is the lwIP default interface, adds the PPP DNS
servers to default DNS server storage. If PPP is not default
interface, then adds DNS servers to interface specific storage.

artokin commented on features/nanostack/mbed-mesh-api/source/NanostackPPPInterface.cpp:
```diff
@@ -0,0 +1,402 @@
+
+void PPPPhy::get_mac_address(uint8_t *mac)
+{
+}
```
Are get_mac_address and set_mac_address left intentionally empty? If so then comment would be nice.

mikaleppanen replied:
Refactored nanostack phy classes so that mac address get/set are no longer part of the base class.

artokin commented on features/netsocket/PPPInterface.cpp:
```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (c) 2018 ARM Limited
```
Please update copyrigth year in all new files

mikaleppanen replied:
Corrected

artokin commented on features/netsocket/PPPInterface.cpp:
```diff
@@ -0,0 +1,193 @@
+
+void PPPInterface::set_as_default()
+{
+#if 0
```
Please add comment why PPPINterface can't be set as default interface

mikaleppanen replied:
Removed the "if 0" and casting to lwip stack.

kjbracey reviewed (approved):
Architecturally, I'm extremely impressed. This pulls together a whole bunch of Mbed OS concepts that have been partially formed since 5.7 and 5.9, and extends them very elegantly to cover this generic PPP engine, 

The whole set-up looks fantastic now, and `OnboardNetworkStack` is beautiful.

My only real concern is the forking out of the PPP core from lwip, and what the plan is to track any upstream lwIP PPP changes in future. Do we have the tooling to do that elegantly? I guess git subtree squash might still manage to handle the file renamings - it's just a merge, ultimately, and that can sometimes...?

VeijoPesonen commented on features/netsocket/NetworkInterface.h:
```diff
@@ -338,6 +339,14 @@ class NetworkInterface: public DNS {
         return 0;
     }
 
+    /** Return pointer to a MeshInterface.
```
PPPInterface

mikaleppanen replied:
Corrected

VeijoPesonen commented on features/netsocket/PPPInterface.h:
```diff
@@ -0,0 +1,197 @@
+     *
+     *  Requires that the network is disconnected
+     *
+     *  @param dhcp     False to disable dhcp (defaults to enabled)
```
Does this actually have a default argument?

mikaleppanen replied:
Removed whole function from PPP interface since dhcp is not valid for ppp. It is set false on stack bring up.

VeijoPesonen commented on features/netsocket/PPPInterface.h:
```diff
@@ -0,0 +1,197 @@
+
+    mbed::FileHandle *_stream;
+    nsapi_ip_stack_t _ip_stack;
+    char _uname[30];
```
Should a define be used instead?

mikaleppanen replied:
Chained user name and password to pointers inside ppp interface class (similar to cellular interface).

VeijoPesonen commented on features/netsocket/ppp/source/auth.c:
```diff
@@ -89,9 +89,9 @@
-#include <netdb.h>
-#include <netinet/in.h>
-#include <arpa/inet.h>
+//#include <netdb.h>
```
Should be removed?

mikaleppanen replied:
Removed.

Commit:
Added PPP directory to astyle ignore and added some spellchecker ignore words

Commit:
Added missing header to wisun interface

0xc0170 reviewed (changes requested):
> [x] Functionality change

Please add release notes section describing this in detail for users

Commit:
Corrected PPP thread stack sizes and network interface

Corrected PPP thread stack size for RZ_A1_EMAC, CYW943012P6EVB_01,
CY8CPROTO_062_4343W, CY8CKIT_062_WIFI_BT and CY8CKIT_062S2_43012
that have special configuration for PPP thread size. Removed
pppInterface() helper call from network interface. It causes binary
compatibility break with precompiled network interface classes. Call
is helper function to check network interface type in case it is
unknown, and is not mandatory or used with PPP.
